### ISSUE TITLE
APIserver is crashing on high load in large clusters

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#26563]
There are two modes of crashing:
- the first one I've seen 5 times today
- the second I've seen exactly once

```
goroutine 3951325 [running]:
runtime.throw(0x2ef0660, 0x15)
        /usr/local/go/src/runtime/panic.go:547 +0x90 fp=0xc8adb6ff68 sp=0xc8adb6ff50
runtime.mapdelete(0x200e720, 0xc857867a10, 0xc8adb70088)
        /usr/local/go/src/runtime/hashmap.go:559 +0x5a fp=0xc8adb6ffc8 sp=0xc8adb6ff68
k8s.io/kubernetes/pkg/api/v1.Convert_api_Pod_To_v1_Pod(0xc903072860, 0xc9031c4150, 0x7f6bb9f96998, 0xc8784fb2c0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/api/v1/conversion.go:514 +0x551 fp=0xc8adb70130 sp=0xc8adb6ffc8
k8s.io/kubernetes/pkg/api/v1.autoConvert_api_PodList_To_v1_PodList(0xc8a4e850e0, 0xc87d278ae0, 0x7f6bb9f96998, 0xc8784fb2c0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/api/v1/conversion_generated.go:4628 +0x31a fp=0xc8adb70238 sp=0xc8adb70130
k8s.io/kubernetes/pkg/api/v1.Convert_api_PodList_To_v1_PodList(0xc8a4e850e0, 0xc87d278ae0, 0x7f6bb9f96998, 0xc8784fb2c0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/api/v1/conversion_generated.go:4639 +0x4b fp=0xc8adb70270 sp=0xc8adb70238
runtime.call64(0xc821791100, 0x342dc10, 0xc89a1cd7d0, 0x2000000030)
        /usr/local/go/src/runtime/asm_amd64.s:473 +0x3e fp=0xc8adb702b8 sp=0xc8adb70270
reflect.Value.call(0x24c3ba0, 0x342dc10, 0x13, 0x2cb4f58, 0x4, 0xc8adb707d8, 0x3, 0x3, 0x0, 0x0, ...)
        /usr/local/go/src/reflect/value.go:435 +0x120d fp=0xc8adb70608 sp=0xc8adb702b8
reflect.Value.Call(0x24c3ba0, 0x342dc10, 0x13, 0xc8adb707d8, 0x3, 0x3, 0x0, 0x0, 0x0)
        /usr/local/go/src/reflect/value.go:303 +0xb1 fp=0xc8adb70668 sp=0xc8adb70608
k8s.io/kubernetes/pkg/conversion.(*Converter).callCustom(0xc8201e2180, 0x2ad7ca0, 0xc8a4e850e0, 0x16, 0x2be8ba0, 0xc87d278ae0, 0x16, 0x24c3ba0, 0x342dc10, 0x13, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/conversion/converter.go:584 +0x4f4 fp=0xc8adb70828 sp=0xc8adb70668
k8s.io/kubernetes/pkg/conversion.(*Converter).convert(0xc8201e2180, 0x2842b40, 0xc8a4e850e0, 0x199, 0x2971ee0, 0xc87d278ae0, 0x199, 0xc8784fb2c0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/conversion/converter.go:627 +0xb76 fp=0xc8adb709e0 sp=0xc8adb70828
k8s.io/kubernetes/pkg/conversion.(*Converter).(k8s.io/kubernetes/pkg/conversion.convert)-fm(0x2842b40, 0xc8a4e850e0, 0x199, 0x2971ee0, 0xc87d278ae0, 0x199, 0xc8784fb2c0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/conversion/converter.go:524 +0x84 fp=0xc8adb70a38 sp=0xc8adb709e0
k8s.io/kubernetes/pkg/conversion.(*Converter).doConversion(0xc8201e2180, 0x2ad7ca0, 0xc8a4e850e0, 0x2be8ba0, 0xc87d278ae0, 0x0, 0xc8c7a7c6a8, 0xc8adb70c78, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/conversion/converter.go:561 +0x45e fp=0xc8adb70bd8 sp=0xc8adb70a38
k8s.io/kubernetes/pkg/conversion.(*Converter).Convert(0xc8201e2180, 0x2ad7ca0, 0xc8a4e850e0, 0x2be8ba0, 0xc87d278ae0, 0x0, 0xc8c7a7c6a8, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/conversion/converter.go:524 +0x28d fp=0xc8adb70ca8 sp=0xc8adb70bd8
k8s.io/kubernetes/pkg/runtime.(*Scheme).UnsafeConvertToVersion(0xc82006a2c0, 0x7f6bba382300, 0xc8a4e850e0, 0x0, 0x0, 0x2cc13b8, 0x2, 0x0, 0x0, 0x0, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/runtime/scheme.go:587 +0x1540 fp=0xc8adb71158 sp=0xc8adb70ca8
k8s.io/kubernetes/pkg/runtime.unsafeObjectConvertor.ConvertToVersion(0xc82006a2c0, 0x7f6bba382300, 0xc8a4e850e0, 0x0, 0x0, 0x2cc13b8, 0x2, 0x0, 0x0, 0x0, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/runtime/helper.go:39 +0x8e fp=0xc8adb711b8 sp=0xc8adb71158
k8s.io/kubernetes/pkg/runtime/serializer/versioning.(*codec).EncodeToStream(0xc872e970e0, 0x7f6bba382300, 0xc8a4e850e0, 0x7f6bb9e66b00, 0xc8c7a7c680, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/runtime/serializer/versioning/versioning.go:279 +0xb19 fp=0xc8adb71438 sp=0xc8adb711b8
k8s.io/kubernetes/pkg/apiserver.writeNegotiated(0x7f6bba386890, 0xc820146150, 0x0, 0x0, 0x2cc13b8, 0x2, 0x7f6bb9f9c258, 0xc8c7a7c680, 0xc8915c6620, 0xc8, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/apiserver.go:445 +0x30e fp=0xc8adb71520 sp=0xc8adb71438
k8s.io/kubernetes/pkg/apiserver.write(0xc8, 0x0, 0x0, 0x2cc13b8, 0x2, 0x7f6bba386890, 0xc820146150, 0x7f6bba382300, 0xc8a4e850e0, 0x7f6bb9f9c258, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/apiserver.go:429 +0xb70 fp=0xc8adb71748 sp=0xc8adb71520
...
```

```
goroutine 244779 [running]:
runtime.throw(0x300f700, 0x21)
        /usr/local/go/src/runtime/panic.go:547 +0x90 fp=0xc873db63a8 sp=0xc873db6390
runtime.mapaccess1_faststr(0x200e720, 0xc89b0196e0, 0xc824b33220, 0x19, 0x4)
        /usr/local/go/src/runtime/hashmap_fast.go:202 +0x5b fp=0xc873db6408 sp=0xc873db63a8
k8s.io/kubernetes/pkg/api/v1.(*ObjectMeta).MarshalTo(0xc873db6820, 0xc8d6ef72c4, 0x563e95, 0x563e95, 0x1d4, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/api/v1/generated.pb.go:4335 +0x1005 fp=0xc873db66e8 sp=0xc873db6408
k8s.io/kubernetes/pkg/api/v1.(*Pod).MarshalTo(0xc873db6800, 0xc8d6ef72c1, 0x563e98, 0x563e98, 0x61a, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/api/v1/generated.pb.go:5035 +0x123 fp=0xc873db6758 sp=0xc873db66e8
k8s.io/kubernetes/pkg/api/v1.(*PodList).MarshalTo(0xc8c78a2de0, 0xc8d6cd0018, 0x78b141, 0x78b141, 0x78b139, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/api/v1/generated.pb.go:5391 +0x30e fp=0xc873db6d18 sp=0xc873db6758
k8s.io/kubernetes/pkg/runtime.(*Unknown).NestedMarshalTo(0xc873db70f0, 0xc8d6cd0004, 0x78b155, 0x78b155, 0x7f91f89d3e00, 0xc8c78a2de0, 0x78b139, 0x410cf8, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/runtime/types_proto.go:47 +0x265 fp=0xc873db6e38 sp=0xc873db6d18
k8s.io/kubernetes/pkg/runtime/serializer/protobuf.(*Serializer).EncodeToStream(0xc8201b40a0, 0x7f91f8ec7068, 0xc8c78a2de0, 0x7f91f8e9d218, 0xc8396ad160, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/runtime/serializer/protobuf/protobuf.go:189 +0x4ac fp=0xc873db7150 sp=0xc873db6e38
k8s.io/kubernetes/pkg/runtime.(*SerializerInfo).EncodeToStream(0xc89e7b1c20, 0x7f91f8ec7068, 0xc8c78a2de0, 0x7f91f8e9d218, 0xc8396ad160, 0x0, 0x0, 0x0, 0x0, 0x0)
        <autogenerated>:30 +0xbe fp=0xc873db71b8 sp=0xc873db7150
k8s.io/kubernetes/pkg/runtime/serializer/versioning.(*codec).EncodeToStream(0xc8783937a0, 0x7f91f8ec7068, 0xc8c78a2de0, 0x7f91f8e9d218, 0xc8396ad160, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/runtime/serializer/versioning/versioning.go:288 +0xc04 fp=0xc873db7438 sp=0xc873db71b8
k8s.io/kubernetes/pkg/apiserver.writeNegotiated(0x7f91f8ecb848, 0xc820400620, 0x0, 0x0, 0x2cc13b8, 0x2, 0x7f91f8e9d078, 0xc8396ad160, 0xc8a8c24620, 0xc8, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/apiserver.go:445 +0x30e fp=0xc873db7520 sp=0xc873db7438
k8s.io/kubernetes/pkg/apiserver.write(0xc8, 0x0, 0x0, 0x2cc13b8, 0x2, 0x7f91f8ecb848, 0xc820400620, 0x7f91f8ec6870, 0xc84aa394a0, 0x7f91f8e9d078, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/apiserver.go:429 +0xb70 fp=0xc873db7748 sp=0xc873db7520
k8s.io/kubernetes/pkg/apiserver.ListResource.func1(0xc876ed7c80, 0xc84aa39200)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:328 +0xe5b fp=0xc873db7b38 sp=0xc873db7748
k8s.io/kubernetes/pkg/apiserver/metrics.InstrumentRouteFunc.func1(0xc876ed7c80, 0xc84aa39200
```

@smarterclayton

**Comment by wojtek-t** [comment]
OK - so I think that the second error is exactly the same as the first one - this is also about annotations in a pod.
I will look into it.

**Comment by lavalamp** [comment]
Thanks for finding! This doesn't look like a scale issue-- it looks like the sort of issue where we need to be finding crashes automatically in test runs...

**Comment by wojtek-t** [comment]
It's not scale issue per-se, but it seems that scale triggers it (i.e. I didn't see that in normal gce suite).

**Comment by dims** [comment]
@wojtek-t the first one seems to be concurrent map write:
https://github.com/golang/go/blob/release-branch.go1.6/src/runtime/hashmap.go#L559

The second one is concurrent read and write:
https://github.com/golang/go/blob/release-branch.go1.6/src/runtime/hashmap_fast.go#L202

**Comment by wojtek-t** [comment]
@dims - we have stack traces, so we basically know what it is...

**Comment by dims** [comment]
@wojtek-t thanks for your patience with my n00b observations:)

**Comment by dims** [comment]
@wojtek-t oops missed the PR (https://github.com/kubernetes/kubernetes/pull/26564/) does that PR fully fix the problem mentioned here? (curious)

**Comment by wojtek-t** [comment]
@dims - that's a good question. I hope so.

**Comment by wojtek-t** [comment]
OK - we had multiple runs where we didn't face this. So I'm closing this as fixed by #26564
